rails_event_store/spec/instrumented_repository_spec.rb

```diff
@@ -0,0 +1,176 @@
+
+        expect(some_repository).to have_received(:delete_stream).with("SomeStream")
+      end
+
+      specify "instruments" do
+        some_repository = double
+        allow(some_repository).to receive(:delete_stream)
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        notification_calls = subscribe_to("delete_stream.repository.rails_event_store")
+
+        instrumented_repository.delete_stream("SomeStream")
+
+        expect(notification_calls).to eq([
+          { stream: "SomeStream" }
+        ])
+      end
+    end
+
+    describe "#has_event?" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+
+        instrumented_repository.has_event?(42)
+
+        expect(some_repository).to have_received(:has_event?).with(42)
+      end
+    end
+
+    describe "#last_stream_event" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+
+        instrumented_repository.last_stream_event("SomeStream")
+
+        expect(some_repository).to have_received(:last_stream_event).with("SomeStream")
+      end
+    end
+
+    describe "#read_event" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+
+        instrumented_repository.read_event(42)
+
+        expect(some_repository).to have_received(:read_event).with(42)
+      end
+
+      specify "instruments" do
+        some_repository = double
+        allow(some_repository).to receive(:read_event)
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        notification_calls = subscribe_to("read_event.repository.rails_event_store")
+
+        instrumented_repository.read_event(42)
+
+        expect(notification_calls).to eq([
+          { event_id: 42 }
+        ])
+      end
+    end
+
+    describe "#read" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        specification = double
+
+        instrumented_repository.read(specification)
+
+        expect(some_repository).to have_received(:read).with(specification)
+      end
+
+      specify "instruments" do
+        some_repository = double
+        allow(some_repository).to receive(:read)
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        notification_calls = subscribe_to("read.repository.rails_event_store")
+        specification = double
+
+        instrumented_repository.read(specification)
+
+        expect(notification_calls).to eq([
+          { specification: specification }
+        ])
+      end
+    end
+
+    def subscribe_to(name)
```

Comment:
As we never unsubscribe, doesn't it lead to memory leaks in all RailsEventStore tests?

Wouldn't these arrays continue receiving payloads for all tests (in this file and other files) which call methods on `RailsEventStore::Clients`. It might be better to verify with temporary subscriptions:

```
callback = lambda {|*args| ... }
ActiveSupport::Notifications.subscribed(callback, /rails_event_store/) do
  ...
end
```

Reply:
```
let(:received) { [] }
around do |spec|
  callback = lambda {|*args| received << args }
  ActiveSupport::Notifications.subscribed(callback, /rails_event_store/) do
    spec.call
  end
end
```

or

```
let(:received) { [] }
around do |spec|
  callback = lambda {|*args| received << args }
  ActiveSupport::Notifications.subscribed(callback, "#{spec.metadata[:subscription_name]}.repository.rails_event_store") do
    spec.call
  end
end

specify "instruments append_to_stream", subscription_name: "append_to_stream" do
  # ...
end
```